### PR TITLE
Allow injecting the user temporarily for direct editing

### LINKED ISSUE #900: Error with opening documents in Nextcloud Mobile App with Groupfolders Advanced Permissions
### Problem
Documents in a Groupfolder cant be open in the Nextcloud Mobile App if with the Groupfolders advanced permissions option "Share" is denied. In the Webbrowser the same documents can be opened without any problems.

### System Details

- Nextcloud 27.1.3
- OnlyofficePlugin 8.2.4
- Onlyoffice 7.5.0.125
- Groupfolders Plugin 15.3.1

### Log Message

{"reqId":"QKEbGPuypE0m2XXXXXXX","level":3,"time":"2023-11-09T15:12:30+00:00","remoteAddr":"xxx.xxx.xxx.xxx","user":"--","app":"onlyoffice","method":"GET","url":"/apps/files/directEditing/36pQZTexHs65H4bfjxW8F2Nysxsxxxxxxxxxxxxxxxxx4mcxgmzYJsF3Szpxz","message":"{\"Exception\":\"OCP\\\\Files\\\\NotFoundException\",\"Message\":\"/2b021b14-9341-103d-90c7-ad6fdc5e7eb0/files/2025d/Path/to/file.xlsx\",\"Code\":0,\"Trace\":[{\"file\":\"/nix/store/i3kxkcvkqkzg425045lwk1c2wlqfk6l1-nextcloud-27.1.3/lib/private/Files/Node/LazyUserFolder.php\",\"line\":74,\"function\":\"get\",\"class\":\"OC\\\\Files\\\\Node\\\\Root\",\"type\":\"->\"},{\"file\":\"/nix/store/i3kxkcvkqkzg425045lwk1c2wlqfk6l1-nextcloud-27.1.3/lib/private/DirectEditing/Manager.php\",\"line\":311,\"function\":\"get\",\"class\":\"OC\\\\Files\\\\Node\\\\LazyUserFolder\",\"type\":\"->\"},{\"file\":\"/nix/store/i3kxkcvkqkzg425045lwk1c2wlqfk6l1-nextcloud-27.1.3/lib/private/DirectEditing/Token.php\",\"line\":50,\"function\":\"getFileForToken\",\"class\":\"OC\\\\DirectEditing\\\\Manager\",\"type\":\"->\"},{\"file\":\"/nix/store/agg6z9mpdbmdcc4h3jr88sda402axymn-nc-app-onlyoffice-8.2.4/lib/DirectEditor.php\",\"line\":208,\"function\":\"getFile\",\"class\":\"OC\\\\DirectEditing\\\\Token\",\"type\":\"->\"},{\"file\":\"/nix/store/i3kxkcvkqkzg425045lwk1c2wlqfk6l1-nextcloud-27.1.3/lib/private/DirectEditing/Manager.php\",\"line\":211,\"function\":\"open\",\"class\":\"OCA\\\\Onlyoffice\\\\DirectEditor\",\"type\":\"->\"},{\"file\":\"/nix/store/i3kxkcvkqkzg425045lwk1c2wlqfk6l1-nextcloud-27.1.3/apps/files/lib/Controller/DirectEditingViewController.php\",\"line\":65,\"function\":\"edit\",\"class\":\"OC\\\\DirectEditing\\\\Manager\",\"type\":\"->\"},{\"file\":\"/nix/store/i3kxkcvkqkzg425045lwk1c2wlqfk6l1-nextcloud-27.1.3/lib/private/AppFramework/Http/Dispatcher.php\",\"line\":230,\"function\":\"edit\",\"class\":\"OCA\\\\Files\\\\Controller\\\\DirectEditingViewController\",\"type\":\"->\"},{\"file\":\"/nix/store/i3kxkcvkqkzg425045lwk1c2wlqfk6l1-nextcloud-27.1.3/lib/private/AppFramework/Http/Dispatcher.php\",\"line\":137,\"function\":\"executeController\",\"class\":\"OC\\\\AppFramework\\\\Http\\\\Dispatcher\",\"type\":\"->\"},{\"file\":\"/nix/store/i3kxkcvkqkzg425045lwk1c2wlqfk6l1-nextcloud-27.1.3/lib/private/AppFramework/App.php\",\"line\":183,\"function\":\"dispatch\",\"class\":\"OC\\\\AppFramework\\\\Http\\\\Dispatcher\",\"type\":\"->\"},{\"file\":\"/nix/store/i3kxkcvkqkzg425045lwk1c2wlqfk6l1-nextcloud-27.1.3/lib/private/Route/Router.php\",\"line\":315,\"function\":\"main\",\"class\":\"OC\\\\AppFramework\\\\App\",\"type\":\"::\"},{\"file\":\"/nix/store/i3kxkcvkqkzg425045lwk1c2wlqfk6l1-nextcloud-27.1.3/lib/base.php\",\"line\":1068,\"function\":\"match\",\"class\":\"OC\\\\Route\\\\Router\",\"type\":\"->\"},{\"file\":\"/nix/store/i3kxkcvkqkzg425045lwk1c2wlqfk6l1-nextcloud-27.1.3/index.php\",\"line\":36,\"function\":\"handleRequest\",\"class\":\"OC\",\"type\":\"::\"}],\"File\":\"/nix/store/i3kxkcvkqkzg425045lwk1c2wlqfk6l1-nextcloud-27.1.3/lib/private/Files/Node/Root.php\",\"Line\":206,\"message\":\"DirectEditor open\",\"CustomMessage\":\"DirectEditor open\"}","userAgent":"Mozilla/5.0 (Android 14) Mobile Nextcloud-android/3.26.0","version":"27.1.3.2"}

### PR DESCRIPTION
**Pull request by juliusknorr**
- Drop usage of incognito mode
- Allow to inject the current user into the session without writing it to the php session itself for direct editing

## Groupfolders

Fix https://github.com/ONLYOFFICE/onlyoffice-nextcloud/issues/900 with https://github.com/ONLYOFFICE/onlyoffice-nextcloud/pull/967

- Setup a groupfolder that has read only ACL set on a child folder (without share permission)
- Try to open a file with a mobile app
- The mobile app uses a webview without user session
- Now the onlyoffice codebase uses getUserFolder()->getById() which is fine as they handle authentication internally, however the getById will fail as groupfolders is not aware of the current user the mountpoint is setup with and assumes that the file is a share (requiring share acls)

### CODE REVIEW

File: apps/files_external/lib/Migration/DummyUserSession.php [hunk 1]
```diff
@@ -30,7 +30,7 @@
 class DummyUserSession implements IUserSession {
 
 	/**
-	 * @var IUser
+	 * @var ?IUser
 	 */
 	private $user;
```
Comment (Fenn-CS) [inline apps/files_external/lib/Migration/DummyUserSession.php:35]:
```suggestion
    private ?IUser $user
```

File: lib/private/User/Session.php [hunk 2]
```diff
@@ -211,6 +211,15 @@ public function setUser($user) {
 		$this->activeUser = $user;
 	}
 
+	/**
+	 * Temporarily set the currently active user without persisting in the session
+	 *
+	 * @param IUser|null $user
+	 */
+	public function setActiveUser(?IUser $user): void {
```
Comment (Fenn-CS) [inline lib/private/User/Session.php:219]:
```suggestion
	public function setVolatileActiveUser(?IUser $user): void {
```

Consideration?? (or a more indicative name) Given what the function actually does?

**Review by nfebe — APPROVED** [review]

File: lib/private/User/Session.php (hunk 2 above)
Comment (juliushaertl) [reply lib/private/User/Session.php]: Good idea, adjusted

**Review by blizzz — APPROVED** [review]

File: lib/private/legacy/OC_User.php [hunk 3]
```diff
@@ -338,7 +339,7 @@ public static function isAdminUser($uid) {
 	 * @return string|false uid or false
 	 */
 	public static function getUser() {
-		$uid = \OC::$server->getSession() ? \OC::$server->getSession()->get('user_id') : null;
+		$uid = Server::get(IUserSession::class)->getUser()?->getUID();
```
Comment (nickvergessen) [inline lib/private/legacy/OC_User.php:342]: This breaks logging with debug when ldap is enabled:
https://github.com/nextcloud/server/pull/44400

**Pull request merged**